### DISCUSSION TITLE
How to mock Session.add() to print SQL to stdout

### DISCUSSION THREAD

**Discussion by ericmjalbert** [sqlalchemy/sqlalchemy#6439] [Usage Questions]
Hi all, I'm trying to get `unittest` to work with my application that uses SQLalchemy and I'm having some problems with getting the SQL statments that the ORM sessions run when doing `session.add()`. 

The way I'm trying to approach the tests is to mock input into the functions and see that SQLalchemy produces the expected SQL statments against the database. For the testing setup I've used [create_mock_engine](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_mock_engine) to get the setup working for DDL operations but I cannot find any way to get a similar approach working for `session.add()`. I'd like to get something setup so that I'd see the `INSERT ...` statements without actually needing to have a database setup.

Here's a minimal example of what I'm trying to get:
```python
from sqlalchemy import Column, String, Integer
from sqlalchemy import create_mock_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session

base = declarative_base()


class User(base):
    __tablename__ = "users"

    id = Column(Integer, primary_key=True)
    name = Column(String)


if __name__ == "__main__":

    def dump(sql, *multiparams, **params):
        """Special function for printing SQL instead of executing it."""
        print(sql.compile(dialect=db.dialect))

    db = create_mock_engine(f"postgresql+psycopg2://", executor=dump)

    # This correctly prints to stdout
    base.metadata.create_all(db)

    # This part doesn't work
    session = Session(db)
    session.add(User(name="John Smith"))
    session.commit()
```

When the above example is run (with `psycopg2` and `sqlalchemy` installed), I can correctly see the sql for creating the table:
```sql
CREATE TABLE users (
        id SERIAL NOT NULL, 
        name VARCHAR, 
        PRIMARY KEY (id)
)
```

However, whenever the session part comes up I get an error [1].

So I'm curious if there is any configurations to change the Session ORM so that I can get the query string that it "would" execute. I've tried searching through the documentation but I can only find examples for the DDL statements or examples that use [Statement object from SQLalchemy Core](https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html#how-do-i-render-sql-expressions-as-strings-possibly-with-bound-parameters-inlined).

I'll gladly accept any help or critiques on this setup 😅 

[1]
<details><summary>Full trace back of error</summary>
<p>

```
Traceback (most recent call last):
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 755, in _connection_for_bind
    elif conn.in_transaction():
AttributeError: 'MockConnection' object has no attribute 'in_transaction'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/main.py", line 30, in <module>
    session.commit()
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1428, in commit
    self._transaction.commit(_to_root=self.future)
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 826, in commit
    self._prepare_impl()
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 805, in _prepare_impl
    self.session.flush()
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3260, in flush
    self._flush(objects)
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3400, in _flush
    transaction.rollback(_capture_exception=True)
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3360, in _flush
    flush_context.execute()
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 453, in execute
    rec.execute(self)
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 627, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 209, in save_obj
    for (
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 370, in _organize_states_for_save
    for state, dict_, mapper, connection in _connections_for_states(
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 1709, in _connections_for_states
    connection = uowtransaction.transaction.connection(base_mapper)
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 623, in connection
    return self._connection_for_bind(bind, execution_options)
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 732, in _connection_for_bind
    conn = self._parent._connection_for_bind(bind, execution_options)
  File "/Users/ericjalbert/Documents/homex/one-off/python-minimal-sqlalchemy-unittest/venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 769, in _connection_for_bind
    conn.close()
AttributeError: 'MockConnection' object has no attribute 'close'
```

</p>
</details>

**Comment by zzzeek** [comment]
hi there -

the session doesn't actually emit any SQL at the point of add(), it emits SQL during [flush](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#flushing).    The unit of work process in that stage is very interactive on the DBAPI cursor object, where it not only emits SQL, it must get responses from all statements as well regarding things like defaults and primary key values generated, rowcounts, etc.  Therefore there is not a straightforward way to "get the SQL" without actually running against a real database.      If you wanted to really run this way, you'd need to create a mock [DBAPI + connection + cursor](https://www.python.org/dev/peps/pep-0249/) that produces the exact responses that would occur from given SQL statements.  This can be done if you were to implement something like a "SQL recorder" that logs the entire interaction of your program with a real database, memoizes all the requests/responses, and can then play them back in a mock context at a later point.

All of that above are things we and others have played with but overall it's way more work than just having a database present for your test suite.   Also, back when we were doing things like that, we didn't have things like Docker or [pifpaf](https://pypi.org/project/pifpaf/) which pretty much solve the problem of being able to have real databases ready to go for test suites.

so to sum up:  possible, with complexity and brittleness, usually not worth it.
